File: addressbook-main/src/main/java/com/addressbook/rest/UniversalController.java
package com.addressbook.rest;

import com.addressbook.LockRecordException;
import com.addressbook.UniversalFieldsDescriptor;
import com.addressbook.dto.*;
import com.addressbook.ignite.GridDAO;
import com.addressbook.model.Alert;
import com.addressbook.model.Breadcrumb;
import com.addressbook.model.User;
import com.addressbook.security.CurrentUser;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.logout.SecurityContextLogoutHandler;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;

@RestController
@RequestMapping(path = "/rest")
public class UniversalController {

    @Autowired
    private CurrentUser currentUser;

    @PostMapping("/getList4UniversalListForm")
    public CompletableFuture<PageDataDto<TableDataDto<?>>> getList(@RequestParam(value = "start") int start,
                                                                @RequestParam(value = "pageSize") int pageSize,
                                                                @RequestParam(value = "sortName") String sortName,
                                                                @RequestParam(value = "sortOrder") String sortOrder,
                                                                @RequestParam(value = "cache") String cache,
                                                                @RequestBody List<FilterDto> filterDto) {
        return CompletableFuture.supplyAsync(() ->
                PageDataDto.<TableDataDto<?>>builder()
                        .data(new TableDataDto<>(GridDAO.selectCachePage(start, pageSize, sortName, sortOrder, filterDto, cache), GridDAO.getTotalDataSize(cache, filterDto)))
                        .fieldDescriptionMap(UniversalFieldsDescriptor.getFieldDescriptionMap(cache)).build());
    }

    @GetMapping("/getContactList")
    public CompletableFuture<PageDataDto<TableDataDto<?>>> getContactList(@RequestParam(value = "personId") String id) {
        return CompletableFuture.supplyAsync(() -> PageDataDto.<TableDataDto<?>>builder().data(new TableDataDto<>(GridDAO.getContactsByPersonId(id))).build());
    }

    @PostMapping(value = "/saveOrCreatePerson")
    public CompletableFuture<PageDataDto<PersonDto>> saveOrCreatePerson(@RequestBody PersonDto personDto) {
        String login = currentUser.getCurrentUser();
        return CompletableFuture.supplyAsync(() -> PageDataDto.<PersonDto>builder().data(GridDAO.createOrUpdatePerson(personDto, login)).build());
    }

    @PostMapping(value = "/saveOrCreateOrganization")
    public CompletableFuture<PageDataDto<OrganizationDto>> saveOrCreateOrganization(@RequestBody OrganizationDto organizationDto) {
        String login = currentUser.getCurrentUser();
        return CompletableFuture.supplyAsync(() -> PageDataDto.<OrganizationDto>builder().data(GridDAO.createOrUpdateOrganization(organizationDto, login)).build());
    }

    @GetMapping(value = "/getBreadcrumbs")
    public CompletableFuture<PageDataDto<List<Breadcrumb>>> getBreadcrumbs(@RequestParam(value = "currentUrl") String url) {
        return CompletableFuture.supplyAsync(() -> PageDataDto.<List<Breadcrumb>>builder().data(GridDAO.readBreadcrumbs(url)).build());
    }

    @GetMapping(value = "/getNextLevelMenus")
    public CompletableFuture<PageDataDto<List<MenuEntryDto>>> getNextLevelMenus(@RequestParam(value = "currentUrl") String url) {
        Collection<? extends GrantedAuthority> authorities = currentUser.getAuthorities();
        return CompletableFuture.supplyAsync(() -> PageDataDto.<List<MenuEntryDto>>builder().data(GridDAO.readNextLevel(url, authorities)).build());
    }

    @PostMapping(value = "/saveOrCreateContacts")
    public CompletableFuture<PageDataDto<List<ContactDto>>> saveOrCreateContacts(@RequestBody List<ContactDto> contactDto, @RequestParam(value = "personId") String personId) {
        String login = currentUser.getCurrentUser();
        return CompletableFuture.supplyAsync(() -> PageDataDto.<List<ContactDto>>builder().data(GridDAO.createOrUpdateContacts(contactDto, login, personId)).build());
    }

    @GetMapping(value = "/lockRecord")
    public PageDataDto<Alert> lockRecord(@RequestParam(value = "type") String type, @RequestParam(value = "id") String id) {
        Alert alert = GridDAO.lockUnlockRecord(type + id, currentUser.getCurrentUser(), true)
                ? Alert.builder().headline("Record locked!").type(Alert.SUCCESS).message("Record id " + id + " locked!").build()
                : Alert.builder().headline("Record was not locked!").type(Alert.WARNING).message("Record id " + id + " was already locked!").build();
        return PageDataDto.<Alert>builder().data(alert).build();
    }

    @GetMapping(value = "/unlockRecord")
    public PageDataDto<Alert> unlockRecord(@RequestParam(value = "type") String type, @RequestParam(value = "id") String id) {
        Alert alert = GridDAO.lockUnlockRecord(type + id, currentUser.getCurrentUser(), false)
                ? Alert.builder().headline("Record unlocked!").type(Alert.SUCCESS).message("Record id " + id + " unlocked!").build()
                : Alert.builder().headline("Record was not unlocked!").type(Alert.WARNING).message("Record id " + id + " was not locked by you!").build();
        return PageDataDto.<Alert>builder().data(alert).build();
    }

    @GetMapping(value = "/logout")
    public String logoutPage(HttpServletRequest request, HttpServletResponse response) {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        if (auth != null) {
            GridDAO.unlockAllRecordsForUser(auth.getName());
            new SecurityContextLogoutHandler().logout(request, response, auth);
        }
        return "redirect:/#/login";
    }

    @ExceptionHandler(Throwable.class)
    public void handleError(HttpServletResponse response, Throwable ex) throws Exception {
        response.setStatus(500);
        Alert alert = Alert.builder().type(Alert.DANGER).headline("Error occurred!").build();
        if (ex.getClass().equals(IllegalArgumentException.class) || ex.getClass().equals(LockRecordException.class)) {
            alert.setMessage(ex.getMessage());
        } else {
            StringWriter errors = new StringWriter();
            ex.printStackTrace(new PrintWriter(errors));
            alert.setMessage(errors.toString());
        }
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.writeValue(response.getWriter(), alert);
    }

    @GetMapping("/getUserInfo")
    public User getUserInfo() {
        User user = GridDAO.getUserByLogin(currentUser.getCurrentUser());
        user.setPassword(null);
        return user;
    }
}

File: addressbook-main/src/main/kotlin/com/addressbook/rest/UniversalController.kt
package com.addressbook.rest

import com.addressbook.LockRecordException
import com.addressbook.UniversalFieldsDescriptor
import com.addressbook.dto.*
import com.addressbook.ignite.GridDAO
import com.addressbook.model.Alert
import com.addressbook.model.Breadcrumb
import com.addressbook.model.User
import com.addressbook.security.CurrentUser
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.logout.SecurityContextLogoutHandler
import org.springframework.web.bind.annotation.*

import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import java.io.PrintWriter
import java.io.StringWriter
import java.util.concurrent.CompletableFuture

@RestController
@RequestMapping(path = ["/rest"])
class UniversalController {

    @Autowired
    var currentUser: CurrentUser? = null

    @PostMapping("/getList4UniversalListForm")
    fun getList(@RequestParam(value = "start") start: Int,
                @RequestParam(value = "pageSize") pageSize: Int,
                @RequestParam(value = "sortName") sortName: String,
                @RequestParam(value = "sortOrder") sortOrder: String,
                @RequestParam(value = "cache") cache: String,
                @RequestBody filterDto: List<FilterDto>): CompletableFuture<PageDataDto<TableDataDto<Any>>> {
        return CompletableFuture.supplyAsync {
            val result = PageDataDto<TableDataDto<Any>>()
            result.data = TableDataDto(GridDAO.selectCachePage(start, pageSize, sortName, sortOrder, filterDto, cache), GridDAO.getTotalDataSize(cache, filterDto))
            result.fieldDescriptionMap = UniversalFieldsDescriptor.getFieldDescriptionMap(cache)
            return@supplyAsync result
        }
    }

    @PostMapping("/getContactList")
    fun getContactList(@RequestParam(value = "personId") id: String): CompletableFuture<PageDataDto<TableDataDto<ContactDto>>> {
        return CompletableFuture.supplyAsync {
            val result = PageDataDto<TableDataDto<ContactDto>>()
            result.data = TableDataDto(GridDAO.getContactsByPersonId(id))
            return@supplyAsync result
        }
    }

    @PostMapping("/saveOrCreatePerson")
    fun saveOrCreatePerson(@RequestBody personDto: PersonDto): CompletableFuture<PageDataDto<PersonDto>> {
        val login = currentUser?.getCurrentUser()!!
        return CompletableFuture.supplyAsync {
            val result = PageDataDto<PersonDto>()
            result.data = GridDAO.createOrUpdatePerson(personDto, login)
            return@supplyAsync result
        }
    }

    @PostMapping("/saveOrCreateOrganization")
    fun saveOrCreateOrganization(@RequestBody organizationDto: OrganizationDto): CompletableFuture<PageDataDto<OrganizationDto>> {
        val login = currentUser?.getCurrentUser()!!
        return CompletableFuture.supplyAsync {
            val result = PageDataDto<OrganizationDto>()
            result.data = GridDAO.createOrUpdateOrganization(organizationDto, login)
            return@supplyAsync result
        }
    }

    @GetMapping("/getBreadcrumbs")
    fun getBreadcrumbs(@RequestParam(value = "currentUrl") url: String): CompletableFuture<PageDataDto<List<Breadcrumb>>> {
        return CompletableFuture.supplyAsync {
            val result = PageDataDto<List<Breadcrumb>>()
            result.data = GridDAO.readBreadcrumbs(url)
            return@supplyAsync result
        }
    }

    @GetMapping("/getNextLevelMenus")
    fun getNextLevelMenus(@RequestParam(value = "currentUrl") url: String): CompletableFuture<PageDataDto<List<MenuEntryDto>>> {
        val authorities = currentUser?.authorities!!
        return CompletableFuture.supplyAsync {
            val result = PageDataDto<List<MenuEntryDto>>()
            result.data = GridDAO.readNextLevel(url, authorities)
            return@supplyAsync result
        }
    }

    @PostMapping("/saveOrCreateContacts")
    fun saveOrCreateContacts(@RequestBody contactDto: List<ContactDto>, @RequestParam(value = "personId") personId: String): CompletableFuture<PageDataDto<List<ContactDto>>> {
        val login = currentUser?.getCurrentUser()!!
        return CompletableFuture.supplyAsync {
            val result = PageDataDto<List<ContactDto>>()
            result.data = GridDAO.createOrUpdateContacts(contactDto, login, personId)
            return@supplyAsync result
        }
    }

    @GetMapping("/lockRecord")
    fun lockRecord(@RequestParam(value = "type") type: String, @RequestParam(value = "id") id: String): PageDataDto<Alert> {
        val result = PageDataDto<Alert>()
        result.data = if (GridDAO.lockUnlockRecord(type + id, currentUser?.getCurrentUser()!!, true))
            Alert("Record locked!", Alert.SUCCESS, Alert.RECORD_PREFIX + id + " locked!")
        else Alert("Record was not locked!", Alert.WARNING, Alert.RECORD_PREFIX + id + " was already locked!")
        return result
    }

    @GetMapping("/unlockRecord")
    fun unlockRecord(@RequestParam(value = "type") type: String, @RequestParam(value = "id") id: String): PageDataDto<Alert> {
        val result = PageDataDto<Alert>()
        result.data = if (GridDAO.lockUnlockRecord(type + id, currentUser?.getCurrentUser()!!, false))
            Alert("Record unlocked!", Alert.SUCCESS, Alert.RECORD_PREFIX + id + " unlocked!")
        else Alert("Record was not unlocked!", Alert.WARNING, Alert.RECORD_PREFIX + id + " was not locked by you!")
        return result
    }

    @GetMapping("/logout")
    fun logoutPage(request: HttpServletRequest, response: HttpServletResponse): String {
        val auth = SecurityContextHolder.getContext().authentication
        if (auth != null) {
            GridDAO.unlockAllRecordsForUser(auth.name)
            SecurityContextLogoutHandler().logout(request, response, auth)
        }
        return "redirect:/#/login"
    }

    @ExceptionHandler(*[Throwable::class])
    fun handleError(response: HttpServletResponse, ex: Throwable) {
        response.status = 500
        val alert = Alert()
        alert.headline = "Error occurred!"
        alert.type = Alert.DANGER
        if (ex.javaClass == IllegalArgumentException::class.java || ex.javaClass == LockRecordException::class.java) {
            alert.message = ex.message
        } else {
            val errors = StringWriter()
            ex.printStackTrace(PrintWriter(errors))
            alert.message = errors.toString()
        }
        val objectMapper = ObjectMapper()
        objectMapper.writeValue(response.writer, alert);
    }

    @GetMapping("/getUserInfo")
    fun getUserInfo(): User? {
        val user = GridDAO.getUserByLogin(currentUser?.getCurrentUser()!!)
        user?.password = null
        return user;
    }
}
